### DISCUSSION TITLE
HTMX Integration

### DISCUSSION THREAD

**Discussion by Hybridhash** [kennyfrc/cami.js#3] [General]
I was looking at the examples provided for instance, Component where there is form is returned from the script. There is one confusion define HTMX attributes ( hx-get, hx-swap, hx- trigger, hx-target etc) as the part of component will work effectively and will be able to alter theDOM  based on hx tags. I believe few examples using HTMX should also included because in any case this library is intended to be used with Hypermedia driven applications.

**Comment by kennyfrc** [comment]
Hi! Yes, that's in the plan. I currently have an application that uses both HTMX and Cami, and I'll extract patterns from it then share in this thread.

**Comment by kennyfrc** [comment]
Here's a simplified / truncated example (I'll put it in the `examples` folder once I figure out how to do a mock server).

This is a rich interactive data table that shows an index of articles. 

Some assumptions:

- This uses ruby's `erb`, a server templating language. Equivalent options would be `jinja` for python or `moustache` for node.js 
- We embed the article data into the `.html.erb` file like so:

```html
<sortable-table-element
    data='<%= h articles.to_json %>'
  ></sortable-table-element>
```

In this example, the ruby code is wrapped around `<%= ... %>`. It converts a ruby hash (or dict if you prefer) into a stringified json, and `h` is a function call that sanitizes the HTML.

In our ReactiveElement constructor, we retrieve the data from the attribute using `this.observableAttr`, like so:

```javascript
      constructor() {
        super();
        this.sortKey = this.observable(null);
        this.sortOrder = this.observable(1); // 1 for ascending, -1 for descending
        this.dataFromAttr = this.observableAttr('data', (data) => {
          return JSON.parse(data).map(item => ({ ...item, selected: false }));
        });
        this.data = this.observable(this.dataFromAttr.value);
      }
   ```
   
Notice that we receive the data as a string, and we need to parse then transform it. `selected: false` is just for the checkbox in our interactive data table.

The template renders a table with interactive features like filter, sort, and select row. At the top, there is a dropdown with name "Bulk Actions", which allows you to bulk delete articles using HTMX. Here, we use `hx-delete` and `hx-vals`. We populate `hx-vals` with a stringified json with `ids` as a key and an `array of ids` as the value.

This is then dispatched to the server, the server deletes the articles for the given ID, then we instruct the client to redirect using the `HX-Redirect` header, which just reloads the screen with the new list of articles.

Here is the shortened code:

```html
<section>
  <sortable-table-element
    data='<%= h articles.to_json %>'
  ></sortable-table-element>
  <script type="module">
    const { html, ReactiveElement } = cami;

    class SortableTableElement extends ReactiveElement {
      constructor() {
        super();
        this.sortKey = this.observable(null);
        this.sortOrder = this.observable(1); // 1 for ascending, -1 for descending
        this.dataFromAttr = this.observableAttr('data', (data) => {
          return JSON.parse(data).map(item => ({ ...item, selected: false }));
        });
        this.data = this.observable(this.dataFromAttr.value);
      }

      // ... rest of the methods ...

      template() {
        return html`
          <nav>
            <h1>Articles</h1>
            <details>
              <summary>Bulk actions</summary>
              <ul>
                <li><a hx-delete="/articles/delete" hx-vals='${JSON.stringify({ ids: [this.data.value.filter(item => item.selected).map(item => item.id)] })}'>Delete</a></li>
              </ul>
            </details>
          </nav>
          <table>
            <thead>
              <tr>
                <th><input type="checkbox" @change=${(e) => this.handleSelectAll(e.target.checked)} /></th>
                <th @click=${() => this.handleSort('title')}>Title</th>
                <th @click=${() => this.handleSort('status')}>Status</th>
                <th @click=${() => this.handleSort('visibility')}>Visibility</th>
              </tr>
              <tr>
                <th></th>
                <th><input type="text" placeholder="Filter by title" @input=${(e) => this.handleFilter('title', e.target.value)} /></th>
                <th><input type="text" placeholder="Filter by status" @input=${(e) => this.handleFilter('status', e.target.value)} /></th>
                <th><input type="text" placeholder="Filter by visibility" @input=${(e) => this.handleFilter('visibility', e.target.value)} /></th>
              </tr>
            </thead>
            <tbody>
              ${this.data.value.map(item => html`
                <tr>
                  <td><input type="checkbox" .checked=${item.selected} @change=${(e) => this.handleSelect(item, e.target.checked)} /></td>
                  <td><a href="/articles/${item.id}">${item.title}</a></td>
                  <td>${item.status}</td>
                  <td>${item.visibility}</td>
                </tr>
              `)}
            </tbody>
          </table>
        `;
      }
    }

    customElements.define('sortable-table-element', SortableTableElement);
  </script>
</section>
```
